Add CS dual accounting

johnwatson484 opened the pull request.

johnwatson484 commented on app/messaging/update-requests-awaiting-manual-ledger-check.js:
```diff
@@ -4,13 +4,13 @@ const enrichPaymentRequests = require('../processing/enrichment')
 
 const updateRequestsAwaitingManualLedgerCheck = async (paymentRequest) => {
-  const orginalPaymentRequest = paymentRequest.paymentRequest
+  const originalPaymentRequest = paymentRequest.paymentRequest
```
Fixing typos in this file

Kirsten1822 replied:
Thanks xd

johnwatson484 commented on app/processing/account-codes/get-codes-for-line.js:
```diff
@@ -0,0 +1,13 @@
+const { isCapital } = require('../is-capital')
+
+const getCodesForLine = (schemeId, lineCode, schemeCode, accountCodeMap) => {
+  if (schemeId === CS) {
+    return accountCodeMap.find(x => x.lineCode === lineCode && (isCapital(schemeCode) ? x.capital : x.revenue))
+  }
```
For CS, we also need to determine whether it's capital or revenue, can't just rely on the line code.

Some line code have a different domestic account for capital vs revenue

johnwatson484 commented on app/processing/transform-payment-request.js:
```diff
@@ -14,17 +14,17 @@ const transformPaymentRequest = async (paymentRequest) => {
   // Check to see if payment request has had a previous payment request.
   // if yes, need to treat as post payment adjustment and calculate Delta which can result in payment request splitting
   const previousPaymentRequests = await getCompletedPaymentRequests(paymentRequest)
+  const sanitizedPaymentRequest = applyDualAccounting(paymentRequest, previousPaymentRequests)
```
Moving dual accounting to before delta calculation.

Reason is that delta is performed at line level which includes grouping by fund code.

If we leave after then we'd have current payment request as `EGF00` and previous all with `DOM10` (for example).  This means we'd always end up creating a negative invoice line for the domestic fund to repay under EU fund code.  Then when dual account is applied we'd end up changing both lines to domestic.  The end result would be fine, but it makes the payment request data more difficult to work with.

Example if we left it after post payment adjustments.

Current payment request -> `EGF00 -> £100`
Previous payment request -> `DOM10` -> £500`

Delta calculation grouped by fund would result in
`EGF00 -> £100`
`DOM10 -> -£500`

After dual accounting would be:
`DOM10 -> £100`
`DOM10 -> -£500`

The result is fine as the customer gets a reduction of -£400 against a domestic fund.  But we've now split that over two lines for no reason.

If however, we apply the domestic fund before delta.  We'd do this.

Current payment request -> `EGF00 -> £100`
Previous payment request -> `DOM10` -> £500`

Apply dual accounting:
Current payment request -> `DOM10 -> £100`
Previous payment request -> `DOM10 -> £500`

Then delta calculation would output
`DOM10 -> -£400`

Same overall result, but much cleaner to read.

Kirsten1822 replied:
I agree, it is much cleaner to read :)

Kirsten1822 commented on app/constants/account-codes/ap.js:
I'm only going to make this comment on the one file, but would it be easier to have the ap, ar admin and ar irregular account codes in order in case we need to find one in future for any particular reason?

johnwatson484 replied:
Makes sense. On it.

johnwatson484 replied:
Done

Kirsten1822 replied:
Thanks :)

Kirsten1822 commented on app/processing/account-codes/map-account-codes.js:
Should this file be renamed to be just map-codes? Since it's in the account-codes file

johnwatson484 replied:
I thought we should keep `account` in there to differentiate it from the files that reference line codes.  So it's clear it's mapping account not line codes.

Kirsten1822 replied:
That's fair enough

Kirsten1822 replied:
Happy to leave as is then

Kirsten1822 commented on app/processing/account-codes/get-line-code.js:
```diff
@@ -0,0 +1,7 @@
+const getLineCode = (description) => {
```
I'm not sure if this name is descriptive enough - I was confusing it with get-codes-for-line

johnwatson484 replied:
Renamed to make it clearer it's getting a code from the description

Kirsten1822 commented on app/processing/account-codes/maps/bps.js:
```diff
@@ -0,0 +1,141 @@
+const { SOS210, SOS404, SOS362, SOS351, SOS363, SOS367, SOS368, SOS369, SOS361, SOS400, SOS402, SOS260, SOS266, SOS410, SOS350, SOS364, SOS415, SOS416, SOS417, SOS420, SOS216, SOS471, SOS472, SOS473, SOS474 } = require('../../../constants/account-codes/ap')
+const { SOS330, SOS465, SOS450, SOS451, SOS452, SOS453, SOS454, SOS455, SOS456, SOS457, SOS458, SOS466, SOS267, SOS459, SOS464, SOS336, SOS460, SOS461, SOS462, SOS463, SOS217 } = require('../../../constants/account-codes/ar-admin')
+const { SOS310, SOS445, SOS430, SOS431, SOS432, SOS433, SOS434, SOS435, SOS436, SOS437, SOS438, SOS446, SOS268, SOS439, SOS444, SOS316, SOS440, SOS441, SOS442, SOS443, SOS218 } = require('../../../constants/account-codes/ar-irregular')
```
Should these be ordered?

johnwatson484 replied:
I put them in order of then being used so you can see the groups of three down each line.

Kirsten1822 commented on app/processing/account-codes/maps/cs.js:
```diff
@@ -0,0 +1,76 @@
+const { SOS710, SOS920, SOS921, SOS922, SOS923, SOS924, SOS925, SOS926, SOS927, SOS928 } = require('../../../constants/account-codes/ap')
+const { SOS750, SOS940, SOS941, SOS942, SOS943, SOS944, SOS945, SOS946, SOS948 } = require('../../../constants/account-codes/ar-admin')
+const { SOS770, SOS960, SOS961, SOS962, SOS963, SOS964, SOS965, SOS966, SOS967, SOS968 } = require('../../../constants/account-codes/ar-irregular')
```
See my comment here: https://github.com/DEFRA/ffc-pay-processing/pull/121/files#r1192436374

johnwatson484 replied:
As before I put them in the order being used, but this one is a bit different as the last code uses an existing one so doesn't quite work.

What do you think?  Either way I'm happy

Kirsten1822 replied:
Happy to leave this as is

Kirsten1822 commented on app/processing/dual-accounting/cs.js:
```diff
@@ -0,0 +1,22 @@
+const { DRD10, EXQ00, DRD00, DRD01, DRD05 } = require('../../constants/domestic-fund-codes')
```
Could these be ordered in someway?

Kirsten1822 commented on app/processing/transform-payment-request.js:
```diff
@@ -14,17 +14,17 @@ const transformPaymentRequest = async (paymentRequest) => {
 
     return deltaPaymentRequests
   }
-  // otherwise first payment request doesn't require delta calculation, so return with dual accounting applied
```
Was this intentionally removed?

Kirsten1822 replied:
Yeah, I'd agree

Kirsten1822 commented on test/unit/processing/account-codes/map-account-codes.test.js:
```diff
@@ -0,0 +1,80 @@
+const { G00 } = require('../../../../app/constants/line-codes')
+const { AP } = require('../../../../app/constants/ledgers')
+const { ADMINISTRATIVE } = require('../../../../app/constants/debt-types')
+const sfiMap = require('../../../../app/processing/account-codes/maps/sfi')
```
Needs a space

Kirsten1822 commented on test/unit/processing/account-codes/map-account-codes.test.js:
```diff
@@ -0,0 +1,80 @@
+    expect(paymentRequest.invoiceLines[0].accountCode).toBe('existing1')
+  })
+
+  test('should get map for scheme', async () => {
```
Any scheme that isn't Manual?

Kirsten1822 commented on test/unit/processing/account-codes/map-account-codes.test.js:
```diff
@@ -0,0 +1,80 @@
+    expect(mockGetLineCode).toHaveBeenCalledWith(paymentRequest.invoiceLines[0].description)
+  })
+
+  test('should get codes for line', async () => {
```
Which codes?

johnwatson484 replied:
Account, updated

Kirsten1822 commented on test/unit/processing/account-codes/select-line-code.test.js:
```diff
@@ -0,0 +1,34 @@
+const { selectLineCode } = require('../../../../app/processing/account-codes/select-line-code')
```
I think this should be below the constants

johnwatson484 replied:
Agreed.  Co-pilot is good at putting them under the last constant which isn't what I always want!!

Kirsten1822 commented on test/unit/processing/dual-accounting/cs.test.js:
```diff
@@ -0,0 +1,80 @@
+const { DRD00, DRD10, EXQ00, DRD01, DRD05 } = require('../../../../app/constants/domestic-fund-codes')
+
+const { applyCSDualAccounting } = require('../../../../app/processing/dual-accounting/cs')
+const capitalSchemes = require('../../../../app/constants/capital-schemes')
```
I think this should be grouped with the other constants

Kirsten1822 commented on test/unit/processing/dual-accounting/get-previous-domestic-fund.test.js:
```diff
@@ -0,0 +1,52 @@
+const domesticFundCodes = require('../../../../app/constants/domestic-fund-codes')
+const { getPreviousDomesticFund } = require('../../../../app/processing/dual-accounting/get-previous-domestic-fund')
```
Space needed here

Kirsten1822 commented on test/unit/processing/dual-accounting/select-domestic-fund-code.test.js:
```diff
@@ -0,0 +1,30 @@
+const { DRD00, DRD01 } = require('../../../../app/constants/domestic-fund-codes')
+const { selectDomesticFundCode } = require('../../../../app/processing/dual-accounting/select-domestic-fund-code')
```
Space needed here

Kirsten1822 commented on test/unit/processing/dual-accounting/select-domestic-fund-code.test.js:
```diff
@@ -0,0 +1,30 @@
+const { DRD00, DRD01 } = require('../../../../app/constants/domestic-fund-codes')
+const { selectDomesticFundCode } = require('../../../../app/processing/dual-accounting/select-domestic-fund-code')
+
+let paymentRequests
```
I think the let should be below the consts and a space needed to separate them

Kirsten1822 commented on test/unit/processing/is-capital.js:
```diff
@@ -0,0 +1,12 @@
+const capitalSchemes = require('../../../app/constants/capital-schemes')
+const { isCapital } = require('../../../app/processing/is-capital')
```
Space needed here

Kirsten1822 commented on app/processing/account-codes/maps/sfi.js:
```diff
@@ -0,0 +1,51 @@
+const { SOS710, SOS920, SOS921, SOS922, SOS923, SOS924, SOS925, SOS926, SOS927 } = require('../../../constants/account-codes/ap')
+const { SOS750, SOS940, SOS941, SOS942, SOS943, SOS944, SOS945, SOS946, SOS947 } = require('../../../constants/account-codes/ar-admin')
+const { SOS770, SOS960, SOS961, SOS962, SOS963, SOS964, SOS965, SOS966, SOS967 } = require('../../../constants/account-codes/ar-irregular')
```
I'm happy to leave these as is

Kirsten1822 commented:
I know we don't necessarily need 100% code coverage, but is there any scenarios in the uncovered code that might need tests?

johnwatson484 commented:
> I know we don't necessarily need 100% code coverage, but is there any scenarios in the uncovered code that might need tests?

The uncovered lines, one is an index file so doesn't have any functionality.

The other is the file that I corrected the spelling it.  That should have tests in it, but don't want to write them all in the scope if this just because I fixed a typo.

It will however be sorted in the other ticket we have to refactor the tests in this repo.

Kirsten1822 reviewed (approved):
Happy to approve :)
